Fix forced inclusion of 'cuda.h' in non-cuda adios2 sources

**Describe the bug**
When building adios2 with CUDA. unrelated source files gets to include `cuda.h` without properly cmake initializing cuda language,

**To Reproduce**
A minimal reproducible example is preferred.
Or Steps to reproduce the behavior:-
1. cmake . -DADIOS_USE_CUDA=ON
2. make /ninja
3. Error messages about not finding cuda.h

```
[1/873] Building CXX object source/adios2/CMakeFiles/adios2_hdf5.dir/toolkit/interop/hdf5/HDF5Common.cpp.o                                                                                                           
FAILED: source/adios2/CMakeFiles/adios2_hdf5.dir/toolkit/interop/hdf5/HDF5Common.cpp.o                                                                                                                               
ccache /usr/bin/clang++-7.0 -DMPICH_SKIP_MPICXX -DOMPI_SKIP_MPICXX -I/home/vicente/Projects/ADIOS2/ADIOS2/source -Isource -march=native -pipe -g -fPIC -std=gnu++11 -MD -MT source/adios2/CMakeFiles/adios2_hdf5.dir/
toolkit/interop/hdf5/HDF5Common.cpp.o -MF source/adios2/CMakeFiles/adios2_hdf5.dir/toolkit/interop/hdf5/HDF5Common.cpp.o.d -o source/adios2/CMakeFiles/adios2_hdf5.dir/toolkit/interop/hdf5/HDF5Common.cpp.o -c /home
/vicente/Projects/ADIOS2/ADIOS2/source/adios2/toolkit/interop/hdf5/HDF5Common.cpp                                                                                                                                    
In file included from /home/vicente/Projects/ADIOS2/ADIOS2/source/adios2/toolkit/interop/hdf5/HDF5Common.cpp:12:                                                                                                     
In file included from /home/vicente/Projects/ADIOS2/ADIOS2/source/adios2/toolkit/interop/hdf5/HDF5Common.tcc:18:                                                                                                     
In file included from /home/vicente/Projects/ADIOS2/ADIOS2/source/adios2/helper/adiosFunctions.h:17:                                                                                                                 
In file included from /home/vicente/Projects/ADIOS2/ADIOS2/source/adios2/helper/adiosMemory.h:262:                                                                                                                   
/home/vicente/Projects/ADIOS2/ADIOS2/source/adios2/helper/adiosMemory.inl:24:10: fatal error: 'cuda.h' file not found
#include <cuda.h>
         ^~~~~~~~
1 error generated.
[2/873] Building CXX object source/adios2/CMakeFiles/adios2_hdf5.dir/engine/hdf5/HDF5WriterP.cpp.o
FAILED: source/adios2/CMakeFiles/adios2_hdf5.dir/engine/hdf5/HDF5WriterP.cpp.o 
ccache /usr/bin/clang++-7.0 -DMPICH_SKIP_MPICXX -DOMPI_SKIP_MPICXX -I/home/vicente/Projects/ADIOS2/ADIOS2/source -Isource -march=native -pipe -g -fPIC -std=gnu++11 -MD -MT source/adios2/CMakeFiles/adios2_hdf5.dir/
engine/hdf5/HDF5WriterP.cpp.o -MF source/adios2/CMakeFiles/adios2_hdf5.dir/engine/hdf5/HDF5WriterP.cpp.o.d -o source/adios2/CMakeFiles/adios2_hdf5.dir/engine/hdf5/HDF5WriterP.cpp.o -c /home/vicente/Projects/ADIOS2
/ADIOS2/source/adios2/engine/hdf5/HDF5WriterP.cpp
In file included from /home/vicente/Projects/ADIOS2/ADIOS2/source/adios2/engine/hdf5/HDF5WriterP.cpp:13:
In file included from /home/vicente/Projects/ADIOS2/ADIOS2/source/adios2/helper/adiosFunctions.h:17:
In file included from /home/vicente/Projects/ADIOS2/ADIOS2/source/adios2/helper/adiosMemory.h:262:
/home/vicente/Projects/ADIOS2/ADIOS2/source/adios2/helper/adiosMemory.inl:24:10: fatal error: 'cuda.h' file not found
#include <cuda.h>
         ^~~~~~~~
1 error generated.
[3/873] Building CXX object source/adios2/CMakeFiles/adios2_hdf5.dir/engine/hdf5/HDF5ReaderP.cpp.o
FAILED: source/adios2/CMakeFiles/adios2_hdf5.dir/engine/hdf5/HDF5ReaderP.cpp.o 
ccache /usr/bin/clang++-7.0 -DMPICH_SKIP_MPICXX -DOMPI_SKIP_MPICXX -I/home/vicente/Projects/ADIOS2/ADIOS2/source -Isource -march=native -pipe -g -fPIC -std=gnu++11 -MD -MT source/adios2/CMakeFiles/adios2_hdf5.dir/
engine/hdf5/HDF5ReaderP.cpp.o -MF source/adios2/CMakeFiles/adios2_hdf5.dir/engine/hdf5/HDF5ReaderP.cpp.o.d -o source/adios2/CMakeFiles/adios2_hdf5.dir/engine/hdf5/HDF5ReaderP.cpp.o -c /home/vicente/Projects/ADIOS2
/ADIOS2/source/adios2/engine/hdf5/HDF5ReaderP.cpp
In file included from /home/vicente/Projects/ADIOS2/ADIOS2/source/adios2/engine/hdf5/HDF5ReaderP.cpp:14:
In file included from /home/vicente/Projects/ADIOS2/ADIOS2/source/adios2/helper/adiosFunctions.h:17:
In file included from /home/vicente/Projects/ADIOS2/ADIOS2/source/adios2/helper/adiosMemory.h:262:
/home/vicente/Projects/ADIOS2/ADIOS2/source/adios2/helper/adiosMemory.inl:24:10: fatal error: 'cuda.h' file not found
#include <cuda.h>
         ^~~~~~~~
1 error generated.
[5/873] Building CXX object source/adios2/CMakeFiles/adios2_core_mpi.dir/engine/insitumpi/InSituMPIReader.cpp.o
ninja: build stopped: subcommand failed.

```
**Expected behavior**
Successful compilation

**Desktop (please complete the following information):**
  - OpenSUSE 15.3 
  - Current master
 
**Additional context**
Add any other context about the problem here.

**Following up**
Was the issue fixed? Please report back.

--- a/source/adios2/core/VariableBase.cpp
+++ b/source/adios2/core/VariableBase.cpp
@@ -23,6 +23,11 @@
 #include "adios2/helper/adiosFunctions.h" //helper::GetTotalSize
 #include "adios2/helper/adiosString.h"
 
+#ifdef ADIOS2_HAVE_CUDA
+#include <cuda.h>
+#include <cuda_runtime.h>
+#endif
+
 namespace adios2
 {
 namespace core

--- a/source/adios2/helper/adiosMemory.cpp
+++ b/source/adios2/helper/adiosMemory.cpp
@@ -15,6 +15,11 @@
 
 #include "adios2/helper/adiosType.h"
 
+#ifdef ADIOS2_HAVE_CUDA
+#include <cuda.h>
+#include <cuda_runtime.h>
+#endif
+
 namespace adios2
 {
 namespace helper
@@ -306,6 +311,13 @@ uint64_t PaddingToAlignOffset(uint64_t offset, uint64_t alignment_size)
     }
     return padSize;
 }
+
+#ifdef ADIOS2_HAVE_CUDA
+void MemcpyGPUToBuffer(void *dst, const char *src, size_t byteCount)
+{
+    cudaMemcpy(dst, src, byteCount, cudaMemcpyDeviceToHost);
+}
+#endif
 
 } // end namespace helper
 } // end namespace adios2

--- a/source/adios2/helper/adiosMemory.h
+++ b/source/adios2/helper/adiosMemory.h
@@ -39,13 +39,18 @@ template <class T>
 void InsertToBuffer(std::vector<char> &buffer, const T *source,
                     const size_t elements = 1) noexcept;
 
+#ifdef ADIOS2_HAVE_CUDA
 /*
  * Copies data from a GPU buffer to a specific location in the adios buffer
  */
-#ifdef ADIOS2_HAVE_CUDA
 template <class T>
 void CopyFromGPUToBuffer(std::vector<char> &buffer, size_t &position,
                          const T *source, const size_t elements = 1) noexcept;
+
+/**
+ * Wrapper around cudaMemcpy needed for isolating CUDA interface dependency
+ */
+void MemcpyGPUToBuffer(void *dst, const char *src, size_t byteCount);
 #endif
 
 /**

--- a/source/adios2/helper/adiosMemory.inl
+++ b/source/adios2/helper/adiosMemory.inl
@@ -20,10 +20,6 @@
 #include <iostream>
 #include <thread>
 /// \endcond
-#ifdef ADIOS2_HAVE_CUDA
-#include <cuda.h>
-#include <cuda_runtime.h>
-#endif
 
 #include "adios2/helper/adiosMath.h"
 #include "adios2/helper/adiosSystem.h"
@@ -84,8 +80,7 @@ void CopyFromGPUToBuffer(std::vector<char> &buffer, size_t &position,
                          const T *source, const size_t elements) noexcept
 {
     const char *src = reinterpret_cast<const char *>(source);
-    cudaMemcpy(buffer.data() + position, src, elements * sizeof(T),
-               cudaMemcpyDeviceToHost);
+    MemcpyGPUToBuffer(buffer.data() + position, src, elements * sizeof(T));
     position += elements * sizeof(T);
 }
 #endif

